### ISSUE TITLE
Configure Proxy per Remote

### ISSUE THREAD

**Issue by monsdar** [conan-io/conan#420]
Currently it's either possible to configure the usage of a proxy globally. It would be better to be able to configure a proxy per remote. That would allow to have conan servers in the LAN and internet at the same time. I could also imagine that there are users that have different proxies for different remotes.

**Comment by memsharded** [comment]
The issue here is that we are not handling the proxies directly in conan, but it is a setting of the underlying ``requests`` library, I have had a look and seems complicated to implement. 

Let me also fully understand the issue. In your setup (company) your internet traffic goes through a proxy, so you need to setup the ``proxies`` field, but then you cannot discover conan servers in the LAN. Wouldn't it be possible to configure the proxy to find such local conan servers?

Also I don't understand multiple proxies for multiple remotes, sounds like a very complex setup. I kind of understand the LAN/proxied-internet issue (basically 1 proxy), but several proxies, not sure. So far, any user has requested this (and there are some using proxies).

Please tell me, and I will have a second look to the possible implementation. Thanks!

**Comment by monsdar** [comment]
You got the setup right - Our company uses a proxy where outgoing traffic is routed through. We also plan on setting up (multiple?) internal conan servers. We probably could set up some no-proxy rules on every client using conan. If it would be easy to implement I could think that conan could profit from a proxy-per-remote configuration.

Reconfiguring the proxy-server itself for conan is rather difficult. It is handled by an external IT-contractor. Brilliant management decision if you ask me ;)

My thought on having multiple proxies at once was just theoretical. I could think about a setup where there is the need of having a proxy per conan server. Though you're right, it's very uncommon and probably easier to integrate workarounds on the user side if this is needed in the future.

**Comment by memsharded** [comment]
Management and brilliant technical decisions in the same sentence! LOL

Ok, I'll have a second look to this issue.

Have you considered using your own proxy server? It is not difficult, and if you are using and managing some servers in your LAN, you can just setup a proxy server in one of your conan LAN servers, and configure the conan ``proxies`` with it. Then configure the proxy server to route to the LAN conan_servers or to the company proxy for external servers. I guess that someone with a little experience could install and configure a Squid server for this purpose in a few minutes.

**Comment by monsdar** [comment]
Using Squid is a good idea, that's what I meant with `probably easier to integrate workarounds on the user side`.

**Comment by memsharded** [comment]
Ok, I see. 
I have checked it again and hacking on the conan client side will be a lot of work, with high risk of breaking things, and at this moment it seems quite a large investment we don't have resources for.

Then yes, I would suggest leaving that as the solution by now. It is clearly a routing problem, and it makes sense to use a solid, well known routing solution (i.e. a proxy server as squid) for it.

To summarize: configure the proxy-server or setup a new intermediate proxy-server in the LAN. Do you agree? Thank you!

**Comment by monsdar** [comment]
I totally agree. Is there some way in Github issues to put issues in some sort of low priority backlog? Else I would close this for now.

**Comment by memsharded** [comment]
I think there is no way to order by priority, we have some label "low priority", but it is not very useful. I suggest closing, and reopening again when it becomes an "important to fix" or a "stopper".
